src/azul/terra.py

```diff
@@ -119,17 +124,22 @@ def parse(cls, spec: str) -> 'TDRSourceSpec':
         # BigQuery (and by extension the TDR) does not allow : or / in dataset names
         # FIXME: Move parsing of prefix to SourceSpec
         #        https://github.com/DataBiosphere/azul/issues/3073
-        service, project, name, prefix = spec.split(':')
-        type, name = name.split('/')
+        service, project, _name, _prefix = spec.split(':')
+        type, name = _name.split('/')
         assert service == 'tdr', service
         if type == cls._type_snapshot:
             is_snapshot = True
         elif type == cls._type_dataset:
             is_snapshot = False
         else:
             assert False, type
-        validate_uuid_prefix(prefix)
-        self = cls(prefix=prefix, project=project, name=name, is_snapshot=is_snapshot)
+        prefix = Prefix.parse(_prefix)
+        if _prefix != str(prefix):
+            spec = ':'.join((service, project, _name, str(prefix)))
```

Comment:
This special case to just make the assertion pass is a bad smell. Parsing and formatting should be the 100% inverse of each other.

Reply:
From this comment it seemed like the delimiter was optional https://github.com/DataBiosphere/azul/pull/2931#discussion_r626111147

> `-` is not allowed to occur in a UUID prefix so we can use it as a delimiter. …:aa-1 means use aa for prefix and a partition prefix of length one. …:aa means to use prefix aa and the default prefix length. …:aa- is not allowed. …: means to use no prefix and the default partition prefix length.

```
Traceback (most recent call last):
  File "<input>", line 1, in <module>
  File "/Users/amar/dev/hca/azul/src/azul/terra.py", line 143, in parse
    assert spec == str(self), (spec, str(self), self)
AssertionError: ('tdr:foo:snapshot/bar:', 'tdr:foo:snapshot/bar:/2', TDRSourceSpec(prefix=Prefix(prefix='', partition_prefix_length=2), project='foo', name='bar', is_snapshot=True))
```
Would you prefer that the assertion would be more loose?